### DISCUSSION TITLE
Is there a way to set Response headers in a server component

### DISCUSSION THREAD

**Discussion by MahmoudBakr22** [vercel/next.js#50116] [App Router]
in server component we only have headers() function which is read-only but I need to set a response header in a server component as this header value depends on some logic and data fetching running in this server component. so it is not possible to use the middleware

**Comment by lucaspiresfernandes** [comment]
Headers and cookies in Server Components are read-only. It's not possible to set headers or cookies in Server Components.

**Reply by MahmoudBakr22** [reply]
so what do you suggest as alternative

**Reply by lucaspiresfernandes** [reply]
There is no alternatives when it comes to Server Components. If you want to set headers, remodel your component so that it makes a call to a Route Handler or to a Server Action.

**Reply by leonbe02** [reply]
That's not really an alternative. A very common use case is to dynamically set `surrogate-key` headers on page responses based on the content that was fetched/rendered. If you use a CDN like Fastly, those [surrogate keys](https://docs.fastly.com/en/guides/working-with-surrogate-keys) are used for programmatic CDN cache invalidation. For example if I had a `/app/products/[slug]/page.js` route and it fetched some product data, I would expect to be able to set a `surrogate-key: <sku-of-the-product>` response header so that if that product were to ever change in my CMS, I could invalidate the CDN cache for just that one page. Prior to the new `/app` router feature in Next.js, you were able to do this via the `getServerSideProps` function since it was passed a reference to the `res` as an argument. Seems like a regression with the new app router

**Reply by MahmoudBakr22** [reply]
@leonbe02 that is easy exactly my case too. In the new App directory this is only available in the middleware but it is not applicable in our case too

**Reply by lucaspiresfernandes** [reply]
I believe this will be added in the future. Currently, the App Router still feels like a beta to me. For the meantime, I would suggest you to create an Route Handler and call that from client when on mount.

**Reply by leonbe02** [reply]
The dynamic response headers are needed on the request for the page itself. Calling an API route from the client-side will only let you apply response headers to that particular API call. In our case, we leverage Fastly (CDN) to cache the server-rendered HTML and we apply custom response headers that tell our CDN how long to cache the HTML and what surrogate keys are used to invalidate it. It doesn't seem like there is currently any way to apply dynamic response headers on that initial page request unless you use middleware which isn't aware of what data was actually fetched during the rendering of the page.

**Reply by MahmoudBakr22** [reply]
I do not know if there is a related issue bit I think we should start one to know if they are planning to add this feature in future or not

**Reply by leonbe02** [reply]
I created one just now: https://github.com/vercel/next.js/issues/50914

**Reply by zhiln** [reply]
this is exactly the issue I'm facing now. I need to tag each page with 'Surrogate-Key' header for Fastly cache purging mechanism. 
Did you find any solution @leonbe02?
I guess the only option now is to switch to /pages router.

**Reply by vaishnavi-jadhav-amla** [reply]
I am also facing the same issue. I need to set a header in one of the server components.

**Reply by salos1982** [reply]
I have the same issue. It is very strange that next.js forbid setting cookies and headers due to streaming. I do not use streaming at all but I need to set headers in response for load balancing and cache.

**Comment by ScreamZ** [comment]
I've a workaround using cookies : But it's dark https://github.com/vercel/next.js/discussions/71415
